File: 109. Convert Sorted List to Binary Search Tree/109. Convert Sorted List to Binary Search Tree.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SortedListToBST(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return new TreeNode(head.val);
        
        ListNode slow = head, fast = head, prev = null;
        while (fast?.next != null) {
            fast = fast.next.next;
            prev = slow;
            slow = slow.next;
        }
        
        TreeNode root = new TreeNode(slow.val);
        if (prev != null) prev.next = null;
        
        if (head != slow) root.left = SortedListToBST(head);
        root.right = SortedListToBST(slow.next);
        
        return root;
    }
}

File: 109. Convert Sorted List to Binary Search Tree/109. Convert Sorted List to Binary Search Tree.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func sortedListToBST(head *ListNode) *TreeNode {
    if head == nil {
        return nil
    }
    if head.Next == nil {
        return &TreeNode{Val: head.Val}
    }
    
    slow, fast := head, head
    var prev *ListNode
    
    for fast != nil && fast.Next != nil {
        fast = fast.Next.Next
        prev = slow
        slow = slow.Next
    }
    
    root := &TreeNode{Val: slow.Val}
    if prev != nil {
        prev.Next = nil
    }
    
    if head != slow {
        root.Left = sortedListToBST(head)
    }
    root.Right = sortedListToBST(slow.Next)
    
    return root
}